### PR TITLE
Sticky table columns and header

### PR DESCRIPTION
**Pull request by CSWinnall**
I've tidied up the code and check for params correctly and removed erroneous classes.

### CODE REVIEW

**Comment by CharlesBrunelleIO** [comment]
First, thanks a lot for your work.
It is, however, better not to close your Pull-Request and reopen a new one each time you want to update your code so we can better keep track of the comments. You can push commits to the same branch, and Github will take care of the rest (updating the PR and notifying the reviewers). This way, we can continue our open conversations and resolve them once finished.

**Review by CharlesBrunelleIO — DISMISSED** [review]
Great 👍  We'll wait on @cortexrd  's approval.

**Comment by CSWinnall** [comment]
Hi Charles 
My apologies.  This is all new to me and I just wanted to make sure everything was right.  I will make sure I don't do that in the future.  Thank you for your advice I appreciate it.
Craig

**Review by cortexrd — COMMENTED** [review]
Works very well!  Even when both are enabled at same time in a give table.
The only thing I've noticed is that, when Bulk Ops are enabled with _stc, the checkboxes are sticky, but the user would also like the first "data" column to be static.  So as an improvement, when Bulk Ops are enabled, stick the first two columns.

For the changes, nothing important, but still would be appreciated:

1. Always leave one empty line between features or new functions.  Ex: lines 5623/24, 5640/41, etc
2. Line 5624 - comment copied but not updated
3. I would prefer we stay with the similar naming convention for keywords entry functions

Ex:

keywords._sth && makeTableHeaderSticky(viewId, keywords, data);
keywords._stc && makeTableColumnSticky(viewId, keywords);

Would be:

keywords._sth && stickyTableHeader(viewId, keywords, data);
keywords._stc && stickyTableColumn(viewId, keywords);

So the letters match the function's name more closely.

Thanks Craig, nice work!
Norm

**Comment by CSWinnall** [comment]
Thanks Norm

I'll make those changes.

**Comment by CSWinnall** [comment]
Hi Norm can you not merge this yet as I am going to work on making sticky columns accept a parameter of num of columns & default to 1 col or 2 based on bulk ops.

**Comment by cortexrd** [comment]
> Hi Norm can you not merge this yet as I am going to work on making sticky columns accept a parameter of num of columns & default to 1 col or 2 based on bulk ops.

Sure, take your time.
Norm

**Comment by CSWinnall** [comment]
I'm having a problem getting the bulkOps checkboxes to stick, as they don't exist until after the keyword runs.  Even if I exclude the getBulkOpsActive and pass in 2 columns the checkboxes still don't stick as they don't exist when the stc code runs not even on a view refresh.  Same goes for the header actually.

image

I'm not sure I want to mess with any of your bulkOps code yet so not sure what to do.

**Comment by CSWinnall** [comment]
I hope this is ok. I think it is finished but I'll be more than happy to make changes.

**Review by cortexrd — APPROVED** [review]
Looks, works nicely - thanks!

**Comment by cortexrd** [comment]
Update:  

1. I found out that both features work as intended upon a fresh page render, but not anymore after the specific view refresh.
2. As a visual improvement, it would be cool to prevent "seeing through" the grid's borders for the sticky columns.  I tried quickly, but without success.  It seems as if the borders are scrolled to the left with the first columns.

**Comment by CSWinnall** [comment]
As to point 2 we have tried for a long time to do that and we haven't found a solution.  We can keep trying though.
I'll take a look at point 1

**Comment by CSWinnall** [comment]
Hi Norm I think I have fixed the issue with the view refresh as to the borders I will keep looking into

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:5659]: What you are doing here is what you think it is. You are only delaying the call to `observerPromise` of 1 sec.
You want to call your promise and then set a timeout.
The question arises: why are you using a promise if you will call `stickTableHeader` a second later anyway? If there is a reason, it would be better to put it in the code, in coding or comments. I firmly believe in descriptive coding (a great read on this is 'Clean Code' by Robert C. Martin).

Also, you should cancel your timeouts instead of using boolean guards. `SetTimeout` returns a unique ID with which you can use `clearTimeout` to cancel its execution.

And, sorry for my long comment, I would suggest leveraging the Promise syntax and writing something like this :
```
let observerPromise = new Promise((resolve) => {
    observeAndExecute(`#bulkOpsControlsDiv-${viewId}`, resolve);
});

observerPromise.then(() => {
        ktl.views.stickTableHeader(viewId, data, numOfRecords, viewHeight);
});
```
It would be easier to read. Fewer JS clutter.

**Comment by cortexrd** [comment]
One more thing... don't forget to remove all console logs.
ND

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:5651]: My comment https://github.com/cortexrd/Knack-Toolkit-Library/pull/148#discussion_r1390412774
still applies.

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:5697]: Some of https://github.com/cortexrd/Knack-Toolkit-Library/pull/148#discussion_r1390412774 can be applied here too.

Comment (CSWinnall) [reply KTL.js:5697]: Thanks Charles.  I will look into this.

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:5727]: You could leverage the function `filter` before `some`. Code like `if (!isEnabled) return false;` will be easier to read.

File: KTL.js
Comment (cortexrd) [reply KTL.js]: Craig, Charles,

When I saw this comment:  `// If observerPromise does not resolve within 1000ms, call callStickTableHeaders`

I had a similar question too:  Why this delay, for what reason, what does it solve as a problem.
It would be nice to explain a bit better in the comment.

Also, I tend to apply Charles recommendations more and more about code simplicity and readability.  I try to make my code almost child-like instead of clever blocks of hard to read/understand code.

I try to use a few extra const/let/var and give them a clear name, then split the code in smaller chunks.

This way, everyone can read it and understand at first glance.

Norm

**Comment by CSWinnall** [comment]
Thanks for the advice I will look at Clean Code and make changes.

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:5761]: @cortexrd  This has a major performance drawback, as I discussed with you. We should avoid this practice as much as possible.

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:9084]: Usage of `const` should as default and `let` only when necessary.

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:9079]: I would specify this is the number of columns to "stick" instead of "display.

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:5698]: After reading `observeAndExecute`,  the 'AndExecute' is misleading. It implies and end to the function call stack. It is more common to name functions like this one `registerObserver` or `registerCallback`. It tells a programmer that the code will be executed more than once and probably forever if there is no 'unregister' .

**Comment by CSWinnall** [comment]
This ended up being quite a lot more complicated than I initially thought.  I have an idea on how to improve the code without using the mutation observer.  I'll give it ago and read up on all the issues that you have told me about so that I can better understand what I'm doing wrong.

Thanks for your suggestions.

**Comment by CSWinnall** [comment]
I ended up going down some rabbit holes with the previous code.  I have taken on board your suggestions the only part I'm not sure about is what to do if the waitSelector has an error. I hope this is a lot simpler.

File: KTL.js
Comment (CharlesBrunelleIO) [inline KTL.js:9022]: `data` and `numOfRecords` are passed only to reject the call of the function. You could stop the execution a lot sooner if you remove that from this function and do that check in the caller function.  Always, the fewer parameters the better.

**Review by CharlesBrunelleIO — COMMENTED** [review]
Good code 👍 

However, you are writing a bit too many comments. Only add a comment to add hidden information or reasons for that specific code. Refrain from adding comments to explain the code itself. Your code is perfectly readable as it is.  Moreover, comments should be as important as the code itself in terms of review and quality, but often they are neglected in reviews and can introduce "errors" misleading the reader.
Some example : 
- `* @param {array} data - The data array` no new information here.
- `// Check if any bulk operation is enabled` and `// if bulkOps are enabled wait to see if the bulkOps...` are on the wrong lines. 
-  `// If keyword has params, update number of columns and background color` you are explaining your very self-explanatory code.

Also, please double-check your `lets` and `consts`. You still have a few extra 'let' that should be `const`.

File: KTL.js
Comment (CSWinnall) [reply KTL.js:9022]: Yes I see that now thank you.

**Pull request merged**
